### PR TITLE
Setting DP to UnsetValue should clear bindings

### LINKED ISSUE #9640: [WASM] DependencyObject.ClearValue not working correctly
### Current behavior

Please see attached and search MainPage.xaml.cs for '!!!' (I added a few comments)

[ClearValue.zip](https://github.com/unoplatform/uno/files/9438210/ClearValue.zip)

### Expected behavior

WASM should work as WindowsAppSDK

### How to reproduce it (as minimally and precisely as possible)

_No response_

### Workaround

_No response_

### Works on UWP/WinUI

Yes

### Environment

Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia

### NuGet package version(s)

_No response_

### Affected platforms

_No response_

### IDE

_No response_

### IDE version

_No response_

### Relevant plugins

_No response_

### Anything else we need to know?

_No response_

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #9640

## PR Type

What kind of change does this PR introduce?

- Bugfix

## What is the current behavior?

Calling `ClearValue` or `SetValue` with `DependencyProperty.UnsetValue` doesn't clear bindings.

## What is the new behavior?

Binding is cleared, as in UWP

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

File: src/Uno.UI.Tests/DependencyProperty/Given_DependencyProperty.cs [hunk 1]
```diff
@@ -1815,6 +1815,17 @@ public void When_DefaultValueOverride()
...
+		[TestMethod]
+		public void ClearValue()
+		{
+			var SUT = new MainPage();
+			Assert.AreEqual(true, SUT.MyCheckBox.IsChecked);
+			Assert.AreEqual(true, SUT.MyModel.IsChecked);
+			SUT.MyButton.RaiseClick();
+			Assert.AreEqual(false, SUT.MyCheckBox.IsChecked);
+			Assert.AreEqual(true, SUT.MyModel.IsChecked);
+		}
```
Comment (jeromelaban) [inline src/Uno.UI.Tests/DependencyProperty/Given_DependencyProperty.cs:1827]: It's probably best to make `DependencyObject`-specific tests, at least one for ClearValue and one for bindings clearing. You could probably move this new test to Button or FrameworkElement related test suites.

File: src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs [hunk 2]
```diff
@@ -562,9 +562,17 @@ internal void SetBindingValue(DependencyPropertyDetails propertyDetails, object
...
 		{
-			SetBindingValue(propertyDetails, args.NewValue);
+			if (isUnsetValue)
+			{
+				propertyDetails.ClearBinding();
+			}
+			else
+			{
+				SetBindingValue(propertyDetails, args.NewValue);
+			}
```
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs:574]: I'm not too sure about removing `SetBindingValue` when `UnsetValue` is used. Is this what UWP/WinUI does?

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs:574]: @jeromelaban My assumption was based on running the following snippet on UWP:

```csharp
			var textBox = new TextBox();
			textBox.SetBinding(TextBox.TextProperty, new Binding());
			var bindingExpr1 = textBox.GetBindingExpression(TextBox.TextProperty);
			textBox.SetValue(TextBox.TextProperty, DependencyProperty.UnsetValue);
			var bindingExpr2 = textBox.GetBindingExpression(TextBox.TextProperty);
```

When running the above code in UWP, I see that `bindingExpr1` is non-null and `bindingExpr2` is null.

That's why I assumed that setting DP to `DependencyProperty.UnsetValue` should clear bindings.

Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs:574]: Indeed it does clear the binding, but indeed, as we're removing the bindings, there's no need to propagate the value. That being said, I'm wondering if there's any sort of propagation that happens to the source anyways. Can you check if there's any callback happening on a `INotifyPropertyChanged` binding on source ?

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs:574]: > Can you check if there's any callback happening on a `INotifyPropertyChanged` binding on source ?

I'm not exactly sure about the scenario you're describing. Do you mean something like this?

```csharp
    public sealed partial class MainPage2 : Page
    {
        public readonly Button MyButton;
        public readonly CheckBox MyCheckBox;
        public readonly Model MyModel;

        public class Model : INotifyPropertyChanged
        {
            private bool isChecked = true;

            public bool IsChecked
            {
                get => isChecked;
                set
                {
                    isChecked = value;
                    if (PropertyChanged != null)
                    {
                        PropertyChanged(this, new PropertyChangedEventArgs(nameof(IsChecked)));
                    }
                }
            }

            public event PropertyChangedEventHandler PropertyChanged;
        }

        public MainPage2()
        {
            MyCheckBox = new CheckBox();
            MyCheckBox.Unchecked += CheckBox_CheckedChanged;
            MyCheckBox.Checked += CheckBox_CheckedChanged;
            MyButton = new Button
            {
                Content = "Click me",
            };
            MyModel = new Model();
            MyModel.PropertyChanged += MyModel_PropertyChanged;
            Content = new StackPanel
            {
                Orientation = Orientation.Horizontal,
                Children =
                {
                    MyCheckBox.Bind(MyModel, nameof(Model.IsChecked)),
                    MyButton.OnClick((s, e) => MyCheckBox.ClearValue(CheckBox.IsCheckedProperty)),
				}
            };
        }

        private void MyModel_PropertyChanged(object sender, PropertyChangedEventArgs e)
        {
            // Check if this is hit?
        }

        private void CheckBox_CheckedChanged(object sender, RoutedEventArgs e) { }
        
    }

    public static class Extensions
    {
        public static TElement Bind<TElement>(this TElement element, object source, string path = null, IValueConverter converter = null, object converterParameter = null, BindingMode mode = BindingMode.TwoWay) where TElement : FrameworkElement
        {
            element.SetBinding(CheckBox.IsCheckedProperty, new Binding
            {
                Converter = converter,
                ConverterParameter = converterParameter ?? element,
                Mode = mode,
                Path = string.IsNullOrEmpty(path) ? null : new PropertyPath(path),
                Source = source ?? element,
            });
            return element;
        }

        public static TElement OnClick<TElement>(this TElement element, RoutedEventHandler handler) where TElement : Button
        {
            element.Click += handler;
            return element;
        }
    }
```

If that's what you meant, then no, `MyModel_PropertyChanged` isn't hit (both UWP and this PR)

**Comment by Youssef1313** [comment]
The `PersonPicture` failure (`VerifyVSMStatesForPhotosAndInitials`) is caused by another inconsistency between Uno and UWP.

I tested the following on UWP:

```xml
<StackPanel>
    <local:MyPersonPicture x:Name="personPicture"/>
    <Button x:Name="MyButton" Content="Click here."></Button>
</StackPanel>
```

```csharp
    public class MyPersonPicture : PersonPicture
    {
        public TextBlock InitialsTextBlock;

        protected override void OnApplyTemplate()
        {
            base.OnApplyTemplate();
            InitialsTextBlock = GetTemplateChild("InitialsTextBlock") as TextBlock;
        }
    }

    public sealed partial class MainPage : Page
    {

        public MainPage()
        {
            this.InitializeComponent();
            this.MyButton.Click += MyButton_Click;
        }

        private void MyButton_Click(object sender, RoutedEventArgs e)
        {
            var exp = personPicture.InitialsTextBlock.GetBindingExpression(TextBlock.FontFamilyProperty);
            // exp is null in UWP
        }
    }
```

However, in Uno, calling `initialsTextBlock.GetBindingExpression(TextBlock.FontFamilyProperty)` in the unit test yields a non-null binding.
I'm not sure, but it looks like this is a difference in how template binding works in UWP vs Uno?

https://github.com/unoplatform/uno/blob/ac0c34e73910ab04499ff29d724f20d9be475fad/src/Uno.UI/UI/Xaml/Controls/PersonPicture/PersonPicture.xaml#L86

What happens is when VisualStateManager.GoToState is called, we clear the setters here:

https://github.com/unoplatform/uno/blob/ac0c34e73910ab04499ff29d724f20d9be475fad/src/Uno.UI/UI/Xaml/VisualStateGroup.cs#L263

which ends with messing up the template binding since we have this setter:

https://github.com/unoplatform/uno/blob/ac0c34e73910ab04499ff29d724f20d9be475fad/src/Uno.UI/UI/Xaml/Controls/PersonPicture/PersonPicture.xaml#L40

I think fixing up the way template binding works is going to be too complex. @jeromelaban Any workarounds you can think of?

**Comment by jeromelaban** [comment]
> I think fixing up the way template binding works is going to be too complex. @jeromelaban Any workarounds you can think of?

`GetBindingExpression` is returning `StaticResource` expressions where it should not, even it is expected to have expressions there (it's used to refresh resources for multiple scenarios). We could probably filter those out, as if I recall properly, they are marked as such.

File: src/Uno.UI.Tests/DependencyProperty/Given_DependencyProperty.cs (hunk 1 above)
Comment (Youssef1313) [reply src/Uno.UI.Tests/DependencyProperty/Given_DependencyProperty.cs]: > It's probably best to make DependencyObject-specific tests, at least one for ClearValue and one for bindings clearing.

@jeromelaban Could you elaborate more on this please? Thanks

**Comment by Youssef1313** [comment]
@jeromelaban This should be ready for review.

File: src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs [hunk 3]
```diff
@@ -562,9 +562,17 @@ internal void SetBindingValue(DependencyPropertyDetails propertyDetails, object
 			return null;
 		}
 
-		private void OnDependencyPropertyChanged(DependencyPropertyDetails propertyDetails, DependencyPropertyChangedEventArgs args)
+		private void OnDependencyPropertyChanged(DependencyPropertyDetails propertyDetails, DependencyPropertyChangedEventArgs args, bool isUnsetValue)
 		{
-			SetBindingValue(propertyDetails, args.NewValue);
+			if (isUnsetValue && args.OldPrecedence == DependencyPropertyValuePrecedences.Local)
```
[the remark this answers is not in the document]
Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/DependencyObjectStore.Binder.cs:567]: > Which tests was the missing local test breaking?

`VerifyVSMStatesForPhotosAndInitials`

---

> I would be interesting to explain why testing to local is useful anyways.

Well, I kept trying different things until I found a working one, so I don't really know :(

**Pull request closed without merging**